composer/loggers/mlflow_logger.py

```diff
@@ -33,6 +36,52 @@
 DEFAULT_MLFLOW_EXPERIMENT_NAME = 'my-mlflow-experiment'
 
 
+class MlflowMonitorProcess(multiprocessing.Process):
+
+    def __init__(self, main_pid, mlflow_run_id, mlflow_tracking_uri):
+        super().__init__()
+        self.main_pid = main_pid
+        self.mlflow_run_id = mlflow_run_id
+        self.mlflow_tracking_uri = mlflow_tracking_uri
+        self.exit_event = multiprocessing.Event()
+        self.crash_event = multiprocessing.Event()
+
+    def handle_sigterm(self, signum, frame):
+        from mlflow import MlflowClient
+        client = MlflowClient(self.mlflow_tracking_uri)
+        if client.get_run(self.mlflow_run_id).info.status == 'RUNNING':
+            # Set the run status as KILLED if SIGTERM is received while the MLflow run is still
+            # in status RUNNING.
+            client.set_terminated(self.mlflow_run_id, status='KILLED')
+
+    def run(self):
+        from mlflow import MlflowClient
+
+        os.setsid()
+        # Register the signal handler in the child process
+        signal.signal(signal.SIGTERM, self.handle_sigterm)
+
+        while not self.exit_event.wait(10):
+            try:
+                # Signal 0 does not kill the process but performs error checking
+                os.kill(self.main_pid, 0)
+            except OSError:
+                client = MlflowClient(self.mlflow_tracking_uri)
+                client.set_terminated(self.mlflow_run_id, status='FAILED')
+                break
+
+        if self.crash_event.is_set():
+            client = MlflowClient(self.mlflow_tracking_uri)
+            client.set_terminated(self.mlflow_run_id, status='FAILED')
+
+    def stop(self):
+        self.exit_event.set()
+
+    def crash(self):
+        self.crash_event.set()
+        self.exit_event.set()
```

Comment:
@chenmoneygithub why do we need two events?

Reply:
So `exit_event ` is for normal exit, and `crash_event` is for crash. If the global exception handler finds an error, it will tell the monitor thread that the main process fails and the monitor thread will set the MLflow run status as `FAILED`.